cumulus_library/base_table_builder.py

```diff
@@ -0,0 +1,49 @@
+""" abstract base for python-based study executors """
+from abc import ABC, abstractmethod
+from typing import final
+
+from cumulus_library.helper import get_progress_bar, query_console_output
+
+
+class BaseTableBuilder(ABC):
+    """Generic base class for python table builders.
+
+    To use a table builder, extend this class exactly once in a new module.
+    See ./studies/vocab or ./studies/core for example usage.
+    """
+
+    def __init__(self):
+        self.queries = []
+        self.display_text = "Building custom tables..."
+
+    @abstractmethod
+    def prepare_queries(self, cursor: object, schema: str):
+        """Main entrypoint for python table builders.
+
+        When completed, prepare_queries should populate self.queries with sql
+        statements to execute. This array will the be read by run queries.
+
+        :param cursor: A PEP-249 compatible cursor
+        :param schema: A schema name
+        :param verbose: toggle for verbose output mode
+        """
+        raise NotImplementedError
+
+    @final
+    def execute_queries(self, cursor: object, schema: str, verbose: bool):
+        """Executes queries set up by a prepare_queries call
+
+        :param cursor: A PEP-249 compatible cursor
+        :param schema: A schema name
+        :param verbose: toggle for verbose output mode
+        """
+        with get_progress_bar(disable=verbose) as progress:
+            task = progress.add_task(
```

Comment:
This is a behavior change, right? Previously there was one progress bar for the whole study? Is this desirable, or just a compromise with how the refactor went? Because I could imagine you having code that sat above all the study's table builders, easily enough.

Reply:
I mean, ultimately who cares whether there's one or two progress bars - I wouldn't worry too much about the effect on the user in this case. Your approach is fine